step/serialization.md

```diff
@@ -1,109 +1,553 @@
+</copy>
+```
+
+Now the filter will reject the incoming serialized data becuase of the presence of `Runnable`: 
+
+```
+Exception in thread "main" java.io.InvalidClassException: filter status: REJECTED
+	at java.base/java.io.ObjectInputStream.filterCheck(ObjectInputStream.java:1360)
+	at java.base/java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:2015)
+	at java.base/java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1872)
+	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2179)
+	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1689)
+	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:495)
+	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:453)
+	at ClassSerializationServer.main(ClassSerializationServer.java:27)
+```
+
+**Extra Credit**: You can further confirm that the filter will only reject `Runnable` by updating `ClassSerializationClient` to send `MessageClass` and see that the code executes successfully. 
+
+## How Records Fixes Serialization
+
+## Serializing a Record
+
+Records are 
+
+```java
+public record MessageRecord(String message) {
+}
+```
+
+**More resources**
 
+* [JEP 395: Records](https://openjdk.java.net/jeps/395)
 
-## Wrap-up
+* [JEP 290: Filter Incoming Serialization Data](https://openjdk.java.net/jeps/290)
 
-**`jlink`** is a tool that can assemble and optimize a set of modules and their dependencies to create a custom, i.e. optimized, run-time Java image. Optimizing the size of a Java runtime is critical when an Java application is embedded with its Java runtime into a container. Reducing the overall container image size will improve the startup time of this container. Moreover, smaller container leads to better resource utilization of the container platform.  It should be mentioned that to use `jlink` the application does not need to be modularized! Moreover, `jlink` is available since JDK 9 so there is no valid reason to not leverage `jlink` today!
+* [Towards Better Serialization](https://cr.openjdk.java.net/~briangoetz/amber/serialization.html) by Brian Goetz
 
-**More ressources**
-* [The `jlink` Command](https://docs.oracle.com/en/java/javase/16/docs/specs/man/jlink.html)
-* [The `jdeps`command](https://docs.oracle.com/en/java/javase/16/docs/specs/man/jdeps.html)
-* [JEP 282: `jlink`](https://openjdk.java.net/jeps/282)
-* [Helidon SE — Custom Runtime Images with `jlink`](https://helidon.io/docs/v2/#/se/guides/37_jlink_image)
-* [Helidon Maven Plugin](https://github.com/oracle/helidon-build-tools/tree/master/helidon-maven-plugin#goal-jlink-image)
-* [Class Data Sharing](https://docs.oracle.com/en/java/javase/16/vm/class-data-sharing.html#GUID-7EAA3411-8CF0-4D19-BD05-DF5E1780AA91)
+* [Serialization and deserialization in Java: explaining the Java deserialize vulnerability](https://snyk.io/blog/serialization-and-deserialization-in-java/) by Brian Vermeer
```

Comment:
A 2020 article on Serialization vulnerability that doesn't mention filter? Please remove!

Reply:
Add
* https://inside.java/2021/04/06/record-serialization-in-practise/
* https://inside.java/2019/11/07/whywehateserialization/